plugins/source/hackernews/resources/services/items/items_fetch.go

```diff
@@ -0,0 +1,158 @@
+package items
+
+import (
+	"container/heap"
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/cloudquery/cloudquery/plugins/source/hackernews/client"
+	"github.com/cloudquery/cloudquery/plugins/source/hackernews/internal/intheap"
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/hermanschaaf/hackernews"
+	"golang.org/x/sync/errgroup"
+)
+
+// fetchItems gets called by the CloudQuery SDK to fetch the data for the `hackernews_items` table. It is passed a context and a
+// schema.ClientMeta interface which is used to access the plugin configuration and the state backend. The
+// resource argument is not used here, as it refers to the parent table, which is not applicable in this case.
+// The res channel is used to send the fetched data back to the CloudQuery SDK.
+//
+// Because this is an incremental table, we must first load the cursor from the state backend, and then update
+// it whenever we've successfully fetched a new item. The cursor is a string, and in this case it is the ID of
+// the last item we've fetched.
+//
+// Incremental tables should guarantee at-least-once delivery. The Hacker News API gives us the current maximum item ID,
+// and we need to fetch all items up to that ID. Once we've fetched all items up to a certain ID, we can safely
+// update the cursor. If we crash before updating the cursor, we will fetch the same items again on the next run.
+//
+// This resolver is different from most in that it needs to concurrently fetch items using its own goroutines. Usually,
+// a resolver with pagination will not have this ability and will simply iterate over all pages.
+func fetchItems(ctx context.Context, meta schema.ClientMeta, _ *schema.Resource, res chan<- any) error {
+	c := meta.(*client.Client)
+	tableName := Items().Name
+	value, err := c.Backend.Get(ctx, tableName, "id")
+	if err != nil {
+		return fmt.Errorf("failed to retrieve state from backend: %w", err)
+	}
+
+	// read the cursor from the state, or default to 0 if it's not set
+	cursor := 0
+	if value != "" {
+		cursor, err = strconv.Atoi(value)
+		if err != nil {
+			return fmt.Errorf("failed to convert cursor to int: %w", err)
+		}
+	}
+	c.Logger().Info().Msg("Found previous cursor, starting from " + strconv.Itoa(cursor))
+
+	// find the max item ID from the Hacker News API
+	work := make(chan int)
+	maxID, err := c.HackerNews.MaxItemID(ctx)
+	if err != nil {
+		return err
+	}
+	c.Logger().Info().Msg("Found max ID, reading up to " + strconv.Itoa(maxID))
+
+	// spin up worker goroutines that will fetch items concurrently by reading ids from
+	// the work channel
+	workers := c.Spec.ItemConcurrency
```

Comment:
I understand this is the most efficient to do this. Another way that might simplify the implementation but is less efficient is to use batches, i.e.

- Start 100 (or whatever) Go routines to get the first batch of items
- If all went well update the cursor and move to the next 100 (or whatever is left)
- If the batch failed set the cursor based on the section of the batch that succeeded

With ⬆️ I don't think we need a heap, only to track the results of the batch

Assuming that API calls take about the same time I'm not sure the performance difference will be that significant